product_unique_default_code/product.py

```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+# © 2016 initOS Gmb
+# License GPL-3.0 or later (http://www.gnu.org/licenses/gpl.html).
+
+from openerp import models, fields, api, _
+from openerp.exceptions import Warning as UserError
+
+
+class product_product(models.Model):
+    _inherit = 'product.product'
+
+    default_code = fields.Char('Internal Reference', copy=False)
+
+    @api.multi
+    @api.constrains('default_code')
+    def _check_unique(self):
+        for obj in self:
+            if 1 < self.search_count(
+                    [('default_code', '=', obj.default_code)]):
+                raise UserError(_("Internal Reference must be unique!"))
```

Comment:
Hi,

As a semantics issue I feel the conditional should read if count > 1.  A unique index on default_code does the same thing although I suppose if duplicates already exist the sql method will not install plus you get a nicer error message here.  

While I didn't test I think there are also possible issues with this constraint.  For example User A creates a product in Company A call PROD.  User B creates a product in Company B called PROD.  Depending on user B's access and record rule setup the constraint either passes or fails?

Also I think it will make variants impossible to use by default as they don't have a default_code at time of creation as standard.  Probably needs a test for that scenario.